Mathlib/Dynamics/TopologicalEntropy/NetEntropy.lean

```diff
@@ -0,0 +1,408 @@
+
+/-!
+# Topological entropy via nets
+We implement Bowen-Dinaburg's definitions of the topological entropy, via nets.
+
+All is stated in the vocabulary of uniform spaces. For compact spaces, the uniform structure
+is canonical, so the topological entropy depends only on the topological structure. This will give
+a clean proof that the topological entropy is a topological invariant of the dynamics.
+
+A notable choice is that we define the topological entropy of a subset `F` of the whole space.
+Usually, one defines the entropy of an invariant subset `F` as the entropy of the restriction of the
+transformation to `F`. We avoid the latter definition as it would involve frequent manipulation of
+subtypes. Our version directly gives a meaning to the topological entropy of a subsystem, and a
+single lemma (`subset_restriction_entropy` in `.Morphism`) will give the equivalence between
+both versions.
+
+Another choice is to give a meaning to the entropy of `∅` (it must be `-∞` to stay coherent) and to
+keep the possibility for the entropy to be infinite. Hence, the entropy takes values in the extended
+reals `[-∞, +∞]`. The consequence is that we use `ℕ∞`, `ℝ≥0∞` and `EReal` numbers.
+
+We relate in this file `CoverEntropy` and `NetEntropy`. This file is downstream of
+`Mathlib.Dynamics.TopologicalEntropy.CoverEntropy` since the submultiplicative argument there
+(specifically `IsDynCoverOf.iterate_le_pow`) is more natural for covers.
+
+## Main definitions
+- `IsDynNetOf`: property that dynamical balls centered on a subset `s` of `F` are disjoint.
+- `netMaxcard`: maximal cardinal of a dynamical net. Takes values in `ℕ∞`.
+- `netEntropyInfEnt`/`netEntropyEnt`: exponential growth of `netMaxcard`. The former is
+defined with a `liminf`, the later with a `limsup`. Take values in `EReal`.
+
+## Implementation notes
+As when using covers, there are two competing definitions `netEntropyInfEnt` and
+`netEntropyEnt` in this file: one uses a `liminf`, the other a `limsup`. When using covers, we chose
+the `limsup` definition as the default. Because of lemmas `coverEntropyInf_eq_iSup_netEntropyInfEnt`
+and `coverEntropy_eq_iSup_netEntropyEnt`, we make the same choice here. Theorems about the
+topological entropy of invariant subsets will be stated using only `coverEntropy`.
+
+## Main results
+- `coverEntropy_eq_iSup_netEntropyEnt`: equality between the notions of topological entropy defined
+with covers and with nets. Has a variant for `coverEntropyÌnf`.
+
+## Tags
+net, entropy
+
+## TODO
+Get versions of the topological entropy on (pseudo-e)metric spaces.
+-/
+
+namespace Dynamics
+
+open Set Uniformity UniformSpace
+
+variable {X : Type*}
+
+/-! ### Dynamical nets -/
+
+/-- Given a subset `F`, an entourage `U` and an integer `n`, a subset `s` of `F` is a
+`(U, n)`-dynamical net of `F` if no two orbits of length `n` of points in `s` shadow each other.-/
+def IsDynNetOf (T : X → X) (F : Set X) (U : Set (X × X)) (n : ℕ) (s : Set X) : Prop :=
```

Comment:
The `Of` seems weird: for the cover, you have a cover of `s`, so it feels right, but there the point is more that you're inside `s`, so `IsDynNetIn` would seem more natural.

Reply:
Nice point. Done.